group/assignment5/main.cpp

```diff
@@ -0,0 +1,194 @@
+#include <iostream>
+#include <cmath>
+#include <vector>
+#include <unordered_map>
+#include "codeu_test_lib.h"
+
+using namespace std;
+
+vector<char> findAlphabetCharacters(const vector<string> &dictionary) {
+    vector<bool> charsInDict((unsigned int)pow(2, CHAR_BIT), false);
+    vector<char> alphabetChars;
+    for (auto word : dictionary) {
+        for (char c : word) {
+            if (!charsInDict[c]) {
+                charsInDict[c] = true;
+                alphabetChars.push_back(c);
+            }
+        }
+    }
+    return alphabetChars;
+}
+
+bool createCharPriorityGraph(vector<string> &dictionary, unordered_map<char, vector<bool>> &charPriorityGraph, unordered_map<char, int> &numOfPriorCharacters) {
+    while (dictionary.size() > 0 && dictionary[0].size() == 0) {
+        dictionary.erase(dictionary.begin());
+    }
+    if (dictionary.size() < 2) {
+        return true;
+    }
+    char prevChar = dictionary[0][0];
+    unordered_map<char, vector<string>> suffixesByCharacters;
+    suffixesByCharacters[prevChar] = vector<string>(0);
+    for (string word : dictionary) {
+        if (word.size() == 0) {
+            continue;
+        }
+        char currentChar = word[0];
+        if (currentChar != prevChar) {
+            if (suffixesByCharacters.find(currentChar) != suffixesByCharacters.end()) { // found a circle in characters - inconsistent dictionary
+                return false;
+            }
+            // mark prevChar as prior to currentChar and add new suffixes vector for current char
+            charPriorityGraph[prevChar][currentChar] = true;
+            suffixesByCharacters[currentChar] = vector<string>(0);
+            numOfPriorCharacters[currentChar]++;
+        }
+        word.erase(0, 1);
+        suffixesByCharacters[currentChar].push_back(word);
+        prevChar = currentChar;
+    }
+    for (auto it : suffixesByCharacters) {
+        vector<string> suffixes = it.second;
+        createCharPriorityGraph(suffixes, charPriorityGraph, numOfPriorCharacters);
+    }
+    return true;
+}
+
+bool findNextOrigin(const unordered_map<char, int> &numOfPriorCharacters, char &origin) {
+    for (auto it : numOfPriorCharacters) {
+        if (it.second == 0) {
+            origin = it.first;
+            return true;
+        }
+    }
+    // no origin -> circle in graph -> dictionary is inconsistent
+    return false;
+}
+
+void updateNumOfPriorCharacters(const unordered_map<char, vector<bool>> &charPriorityGraph, unordered_map<char, int> &numOfPriorCharacters, const char origin) {
+    const vector<bool> children = charPriorityGraph.at(origin);
```

Comment:
I didn't know that instead of operator[] you can use function at(int pos), which will check if index is out of bounds before it try to access the element at specified position. :) thanks, now I know the safe way to access the element of vector :)